File: src/main/java/com/github/tianchenghang/tools/WebSearchTool.java
package com.github.tianchenghang.tools;

import cn.hutool.http.HttpUtil;
import cn.hutool.json.JSONObject;
import cn.hutool.json.JSONUtil;
import java.util.HashMap;
import java.util.stream.Collectors;
import org.springframework.ai.tool.annotation.Tool;
import org.springframework.ai.tool.annotation.ToolParam;

public class WebSearchTool {

  private static final String SEARCH_API_URL = "https://www.searchapi.io/api/v1/search";

  private final String apiKey;

  public WebSearchTool(String apiKey) {
    this.apiKey = apiKey;
  }

  @Tool(description = "Search for information from web search engine")
  public String searchWeb(@ToolParam(description = "Search query keyword") String query) {
    var paramMap = new HashMap<String, Object>();
    paramMap.put("q", query);
    paramMap.put("api_key", apiKey);
    paramMap.put("engine", "google");
    try {
      var response = HttpUtil.get(SEARCH_API_URL, paramMap);
      // Extract top 5 results
      var jsonObject = JSONUtil.parseObj(response);
      var organicResults = jsonObject.getJSONArray("organic_results");
      var objects = organicResults.subList(0, 5);
      // Format search results as a string
      var result =
          objects.stream()
              .map(
                  obj -> {
                    var tmpJsonObj = (JSONObject) obj;
                    return tmpJsonObj.toString();
                  })
              .collect(Collectors.joining(","));
      return result;
    } catch (Exception e) {
      return "Error searching web: " + e.getMessage();
    }
  }
}

File: server/src/tools/web-search-tool.ts
import axios from 'axios';
import { StructuredTool } from '@langchain/core/tools';
import { z } from 'zod';

export class WebSearchTool extends StructuredTool {
  name = WebSearchTool.name;
  description = 'Search for information from web search engine';
  schema = z.object({
    query: z.string().describe('Search query keyword'),
  });

  private readonly searchApiUrl = 'https://www.searchapi.io/api/v1/search';
  private readonly apiKey: string;

  constructor(apiKey?: string) {
    super();
    this.apiKey = apiKey ?? process.env.SEARCH_API_KEY ?? '';
  }

  async _call(args: { query: string }): Promise<string> {
    const { query } = args;
    if (!query) {
      return 'Query parameter is required';
    }
    if (!this.apiKey) {
      return 'Api key is required';
    }
    try {
      const response = await axios.get<{ organic_results: string[] }>(
        this.searchApiUrl,
        {
          params: {
            q: query,
            api_key: this.apiKey,
            engine: 'google',
          },
          timeout: 30_000,
        },
      );
      const { organic_results: originalResults } = response.data;
      const topResults = originalResults.slice(0, 5);
      return JSON.stringify(topResults, null, 2);
    } catch (err) {
      console.error('Searching web error:', err);
      return 'Searching web error';
    }
  }
}
